loky/backend/_thread_pool_limiters.py

```diff
@@ -0,0 +1,492 @@
+
+    cls_thread_locals = threading.local()
+
+    def __init__(self):
+        self._load()
+
+    def _load(self):
+        if sys.platform == "darwin":
+            self._modules = self._find_modules_with_clibs_dyld()
+        elif sys.platform == "win32":
+            self._modules = self._find_modules_with_enum_process_module_ex()
+        else:
+            self._modules = self._find_modules_with_clibs_dl_iterate_phdr()
+
+    def _unload(self):
+        del self._modules
+
+    def starts_with_any(self, library_basename, filename_prefixes):
+        return any([library_basename.startswith(prefix)
+                    for prefix in filename_prefixes])
+
+    def _is_supported_implementation(self, module_path):
+        module_name = os.path.basename(module_path).lower()
+        for name, info in SUPPORTED_IMPLEMENTATION.items():
+            if self.starts_with_any(module_name, info['filename_prefixes']):
+                return name, info['library']
+        return None, None
+
+    def _make_module(self, name, module_path, library):
+        module_path = os.path.normpath(module_path)
+        module = ctypes.CDLL(module_path)
+        set_func = getattr(module, LIB_TO_FUNC_MAP[library]['set'],
+                           lambda n_thread: None)
+        get_func = getattr(module, LIB_TO_FUNC_MAP[library]['get'],
+                           lambda: None)
+        return dict(name=name, library=library, module_path=module_path,
+                    module=module, version=self.get_version(module, library),
+                    set=set_func, get=get_func)
+
+    def get_limit(self, name, libname, limits):
+        if name in limits:
+            return limits[name]
+        if libname in limits:
+            return limits[libname]
+        return None
+
+    def set_thread_limits(self, limits=None, apis=None):
+        """Limit maximal number of threads used by supported C-libraries.
+
+        Return a dict of pairs {clib: bool} where bool is True when clib can be
+        dynamically scaled with this function.
+        """
+        if isinstance(limits, int) or limits is None:
+            if apis in ("all", None):
+                apis = LIB_TO_FUNC_MAP.keys()
+            elif apis == "blas":
+                apis = ("openblas", "mkl")
+            elif apis == "openmp":
+                apis = ("openmp",)
+            else:
+                raise ValueError("apis must be either 'all', 'blas' or "
+                                 "'openmp'. Got {} instead.".format(apis))
+            limits = {api: limits for api in apis}
+        if isinstance(limits, list):
+            limits = {module['name']: module['n_thread'] for module in limits}
+
+        if not isinstance(limits, dict):
+            raise TypeError("limits must either be an int, a dict or None. "
+                            "Got {} instead".format(type(limits)))
+
+        report_threadpool_size = []
+        self._load()
```

Comment:
Instead of `self._load()` and `_unload()` that mutate a `self._modules` attribute, what about:

```python
modules = self._load_modules(user_api=name_or_None, name=implementation_name_or_None, fetch_version=False)
```
and then never store the list of modules as an attribute and therefore we would never need to have an `_unload()` method.

Putting the filtering logic (by user API name or implementation name) inside the stateless `_load_modules` method would further make it possible to skip the ctypes loading of libraries we are not interested in (e.g. no openmp loading when we want to limit the threads of BLAS runtimes).

Reply:
ok